frame/alliance/src/migration.rs

```diff
@@ -41,9 +41,15 @@ pub fn migrate<T: Config<I>, I: 'static>() -> Weight {
 pub struct Migration<T, I = ()>(PhantomData<(T, I)>);
 
 impl<T: Config<I>, I: 'static> OnRuntimeUpgrade for Migration<T, I> {
+	type PreStateDigest = ();
```

Comment:
Can't we make this type be two different generics that `FullCodec` in each function, as the issue recommended? Having to define this type also increases the footprint of this PR a lot. 

Something like:

```
fn pre_upgrade<I: Encode>() -> Result<I, ..> {}
fn post_upgrade<I: Decode>(input: I) {}
```

and we encode and pass the data between the hooks automatically. 

This is more or less what we do now with `OnRuntimeUpgradeHelpersExt`, and if we do this we can remove this extension trait + make sure the state root is not changing in the pre/post migrate hooks.

Reply:
I think we can't do that, the type parameters of a generic function are instantiated when calling the function not when we implement traits (defining the function).

Using concrete types like: 
```
fn pre_upgrade() -> Result<Vec<u8>, ..> {}
fn post_upgrade(input: Vec<u8>) {}
```
may work though, but requiring to do encode/decode within `pre_upgrade/post_upgrade `

Reply:
> I think we can't do that, the type parameters of a generic function are instantiated when calling the function not when we implement traits (defining the function).

Not sure if this is an issue, can you give it a try? 

All in all, I don't want to pollute the code too much for this feature. In rare cases we might need to pass data between the two hooks. Let's just make it `Vec<u8>` (or `T: Codec`) and move on without making a big change.

Making sure that the hooks don't alter storage is a good change though, let's keep that part.

Reply:
Yes, I had tried,
```
impl OnRuntimeUpgrade for SomeType {
    fn pre_upgrade<u32>() -> Result<u32, &'static str> {
        ...
    }
}
```
the compiler will recognize `u32` as a type parameter instead of a concrete type. 

I will try to it `Vec<u8>`.

Reply:
> I will try to it Vec<u8>.

I really made progress on this approach, except for one (perhaps final) issue that `OnRuntimeUpgrade for Tuple` failed due to `Codec` did not implement for tuple with elements more than 18. This may be related to https://github.com/paritytech/parity-scale-codec/blob/46859b803c94f091e3ca644d8468bfadfbd2c539/src/max_encoded_len.rs#L79-L81